xarray/core/dataarray.py

```diff
@@ -244,6 +240,16 @@ def _replace_maybe_drop_dims(self, variable, name=__default):
                                  if set(v.dims) <= allowed_dims)
         return self._replace(variable, coords, name)
 
+    def _replace_indexes(self, indexes):
+        obj = self
+        for dim, idx in iteritems(indexes):
+            if idx.name is None:
+                idx.name = dim + "_unnamed_level"
+            obj = obj.rename({dim: idx.name})
+            new_coord = Coordinate(idx.name, idx)
+            obj = obj._replace(coords={idx.name: new_coord})
```

Comment:
It's better to accumulate all the new coordinates in a dictionary, and then call `_replace` once outside the loop.